usbSerialForAndroid/src/main/java/com/hoho/android/usbserial/util/SerialInputOutputManager.java

```diff
@@ -155,81 +157,141 @@ public void writeAsync(byte[] data) {
+                Log.w(TAG, "Exception in onRunError: " + t.getMessage(), t);
             }
-            mState = State.RUNNING;
         }
-        Log.i(TAG, "Running ...");
+    }
+
+    /**
+     * Set the thread priority
+     */
+    private void setThreadPriority() {
+        if (mThreadPriority != Process.THREAD_PRIORITY_DEFAULT) {
+            Process.setThreadPriority(mThreadPriority);
+        }
+    }
+
+    /**
+     * Continuously services the read buffers until {@link #stop()} is called, or until a driver exception is
+     * raised.
+     */
+    private void runRead() {
+        Log.i(TAG, "runRead running ...");
         try {
-            if(mThreadPriority != Process.THREAD_PRIORITY_DEFAULT)
-                Process.setThreadPriority(mThreadPriority);
-            while (true) {
-                if (getState() != State.RUNNING) {
-                    Log.i(TAG, "Stopping mState=" + getState());
-                    break;
-                }
-                step();
-            }
+            setThreadPriority();
+            mStartuplatch.countDown();
+            do {
+                stepRead();
+            } while (isStillRunning());
+            Log.i(TAG, "runRead: Stopping mState=" + getState());
         } catch (Throwable e) {
-            if(mSerialPort.isOpen()) {
-                Log.w(TAG, "Run ending due to exception: " + e.getMessage(), e);
+            if (Thread.currentThread().isInterrupted()) {
+                Log.w(TAG, "runRead: interrupted");
+            } else if(mSerialPort.isOpen()) {
+                Log.w(TAG, "runRead ending due to exception: " + e.getMessage(), e);
             } else {
-                Log.i(TAG, "Socket closed");
+                Log.i(TAG, "runRead: Socket closed");
             }
-            final Listener listener = getListener();
-            if (listener != null) {
-                try {
-                    if (e instanceof Exception) {
-                        listener.onRunError((Exception) e);
-                    } else {
-                        listener.onRunError(new Exception(e));
-                    }
-                } catch (Throwable t) {
-                    Log.w(TAG, "Exception in onRunError: " + t.getMessage(), t);
+            notifyErrorListener(e);
+        } finally {
+            if (!mState.compareAndSet(State.RUNNING, State.STOPPING)) {
+                if (mState.compareAndSet(State.STOPPING, State.STOPPED)) {
+                    Log.i(TAG, "runRead: Stopped mState=" + getState());
                 }
             }
+            mShutdownlatch.countDown();
+        }
+    }
+
+    /**
+     * Continuously services the write buffers until {@link #stop()} is called, or until a driver exception is
+     * raised.
+     */
+    private void runWrite() {
+        Log.i(TAG, "runWrite running ...");
+        try {
+            setThreadPriority();
+            mStartuplatch.countDown();
+            do {
+                stepWrite();
```

Comment:
busy loop. should only wake up on data or other thread terminating

Reply:
fixed